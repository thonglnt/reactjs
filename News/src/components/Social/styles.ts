// lib
import styled from 'styled-components'

// themes
import { theme } from '@themes'

export const IconStyled = styled.i`
  color: ${(props) => props.color};
  padding-left: 10px;

  &:hover {
    color: ${theme.Colors.crayola};
  }
`

export const ButtonStyled = styled.button`
  background-color: ${(props) => props.color};
  color: white;
  border: none;
  padding: 5px 10px;
  margin-left: 5px;
  border-radius: 3px;

  :hover {
    background-color: ${theme.Colors.crayola};
    color: ${theme.Colors.white};
  }
`

export const IconTagStyled = styled.i`
  margin-right: 3px;
`
