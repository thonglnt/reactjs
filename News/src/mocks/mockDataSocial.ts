// types
import { ISocialButton, ISocialIcon } from 'types/social'

export const mockDataSocialIcon: ISocialIcon[] = [
  { className: 'fa fa-facebook-f', href: '' },
  { className: 'fa fa-twitter', href: '' },
  { className: 'fa fa-pinterest-p', href: '' },
  { className: 'fa fa-vimeo', href: '' },
  { className: 'fa fa-youtube', href: '' },
]

export const mockDataSocialButton: ISocialButton[] = [
  { className: 'fa fa-facebook-f', href: '', nameButton: 'Facebook', backgroundColor: '#3b5998' },
  { className: 'fa fa-twitter', href: '', nameButton: 'Twitter', backgroundColor: '#1da1f2' },
  { className: 'fa fa-google-plus', href: '', nameButton: 'Gooogle+', backgroundColor: '#dd4b39' },
  { className: 'fa fa-pinterest-p', href: '', nameButton: 'Pinterest', backgroundColor: '#bd081c' },
]
