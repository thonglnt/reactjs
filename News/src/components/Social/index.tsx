// lib
import { memo } from 'react'
import isEqual from 'react-fast-compare'

// themes
import { theme } from '@themes'

// types
import { ISocialButton, ISocialIcon } from 'types/social'

// styles
import { IconTagStyled, ButtonStyled, IconStyled } from './styles'

export type SocialProps = {
  icons?: ISocialIcon[]
  buttons?: ISocialButton[]
  size?: 'fa-lg' | 'fa-sm'
  colorIcon: string
}

// deep compare in memo
const areEqual = (prevData: SocialProps, nextData: SocialProps) => {
  return (
    isEqual(prevData.icons, nextData.icons) &&
    isEqual(prevData.buttons, nextData.buttons) &&
    isEqual(prevData.size, nextData.size) &&
    isEqual(prevData.colorIcon, nextData.colorIcon)
  )
}

const Social = ({
  icons = [],
  buttons = [],
  size = 'fa-sm',
  colorIcon = theme.Colors.chineseSilver,
}: SocialProps) => {
  return (
    <div>
      {icons.length > 0
        ? icons.map((item) => (
            <a href={item.href}>
              <IconStyled
                data-testid='colorIcon'
                className={`${item.className} ${size}`}
                color={colorIcon}
              />
            </a>
          ))
        : buttons &&
          buttons.map((item) => (
            <ButtonStyled data-testid='colorButton' color={item.backgroundColor}>
              <IconTagStyled className={`${item.className} ${size}`} />
              {item.nameButton}
            </ButtonStyled>
          ))}
    </div>
  )
}

export default memo(Social, areEqual)
