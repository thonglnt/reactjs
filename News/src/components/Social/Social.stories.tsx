// lib
import { Story } from '@storybook/react'

// Components
import Social, { SocialProps } from '.'

// mocks
import { mockDataSocialButton, mockDataSocialIcon } from '@mocks/mockDataSocial'

// themes
import { theme } from '@themes'

export default {
  title: 'Components/Social',
  component: Social,
}

const Template: Story<SocialProps> = (args) => <Social {...args} />

export const SocialIcon = Template.bind({})
SocialIcon.args = {
  icons: mockDataSocialIcon,
  size: 'fa-lg',
  colorIcon: theme.Colors.chineseSilver,
}

export const SocialButton = Template.bind({})
SocialButton.args = {
  buttons: mockDataSocialButton,
  colorIcon: theme.Colors.white,
}
