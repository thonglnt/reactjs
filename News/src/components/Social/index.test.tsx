// lib
import { render, screen } from '@testing-library/react'

// mocks
import { mockDataSocialButton, mockDataSocialIcon } from '@mocks/mockDataSocial'

// component
import Social, { SocialProps } from '.'

// themes
import { theme } from '@themes'

describe('Test Social component', () => {
  const defaultProps: SocialProps = {
    icons: mockDataSocialIcon,
    size: 'fa-lg',
    colorIcon: theme.Colors.black,
  }

  const customProps: SocialProps = {
    buttons: mockDataSocialButton,
    colorIcon: theme.Colors.chineseBlue,
  }

  test('It should match DOM Snapshot', () => {
    const { container } = render(<Social {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('Component Social it should match Default Props', () => {
    render(<Social {...defaultProps} />)

    expect(screen.getAllByTestId('colorIcon')[0]).toHaveAttribute('color', theme.Colors.black)
  })

  test('Component Social it should match Custom Props', () => {
    render(<Social {...customProps} />)

    expect(screen.getAllByTestId('colorButton')[0]).toHaveAttribute(
      'color',
      theme.Colors.chineseBlue,
    )
  })
})
