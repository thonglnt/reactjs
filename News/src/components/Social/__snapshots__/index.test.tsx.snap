// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Social component It should match DOM Snapshot 1`] = `
<div>
  <div>
    <a
      href=""
    >
      <i
        class="sc-bcXHqe yIwqc fa fa-facebook-f fa-lg"
        color="#000"
        data-testid="colorIcon"
      />
    </a>
    <a
      href=""
    >
      <i
        class="sc-bcXHqe yIwqc fa fa-twitter fa-lg"
        color="#000"
        data-testid="colorIcon"
      />
    </a>
    <a
      href=""
    >
      <i
        class="sc-bcXHqe yIwqc fa fa-pinterest-p fa-lg"
        color="#000"
        data-testid="colorIcon"
      />
    </a>
    <a
      href=""
    >
      <i
        class="sc-bcXHqe yIwqc fa fa-vimeo fa-lg"
        color="#000"
        data-testid="colorIcon"
      />
    </a>
    <a
      href=""
    >
      <i
        class="sc-bcXHqe yIwqc fa fa-youtube fa-lg"
        color="#000"
        data-testid="colorIcon"
      />
    </a>
  </div>
</div>
`;
